# Task 4.8 - Grouping data and aggregating variables

1. Import Libraries
2. Import data
3. Grouping the data
4. Aggregating data
5. Aggregating data with transform
6. Task 4.8
7. Export data

# 1. Import Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 2. Import data

In [2]:
# Defining path
path=r'C:\Users\ponni\Achievement 4'

In [3]:
# Import merged data
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_update.pkl'))

In [4]:
# Create a subset
df=ords_prods_merge[:1000000]

In [5]:
# Check the dimension
df.shape

(1000000, 18)

In [6]:
df.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                   999.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest day        busiest_days busiest_period_of_day  
0  Regularly busy  Regulary busy days           Most orders  
1  Regularly busy        Slowest days        Average orders  
2  Regularly busy        Slowest days           Most orders  
3      Least busy        Slowest days        Average orders  
4      Least busy        Slowest days           Most orders

# 3. Grouping the data

In [7]:
df.groupby('product_name')

# 4. Aggregating data

In [8]:
# performing single agregation - calculating mean
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [9]:
# performing multiple aggregation
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

# 5. Aggregating data with transform

In [10]:
# Creating loyalty flag

In [11]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [12]:
ords_prods_merge.head(15)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    2539329        1             1          2                  8   
1    2398795        1             2          3                  7   
2     473747        1             3          3                 12   
3    2254736        1             4          4                  7   
4     431534        1             5          4                 15   
5    3367565        1             6          2                  7   
6     550135        1             7          1                  9   
7    3108588        1             8          1                 14   
8    2295261        1             9          1                 16   
9    2550362        1            10          4                  8   
10   2968173       15            15          1                  9   
11   1870022       15            17          2                 16   
12   1911383       15            18          2                 11   
13   2715276       15            21          1                  9   
14    487368       15            22          1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    999.0         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  Mid-range product   
1          Soda        77              7     9.0   both  Mid-range product   
2          Soda        77              7     9.0   both  Mid-range product   
3          Soda        77              7     9.0   both  Mid-range product   
4          Soda        77              7     9.0   both  Mid-range product   
5          Soda        77              7     9.0   both  Mid-range product   
6          Soda        77              7     9.0   both  Mid-range product   
7          Soda        77              7     9.0   both  Mid-range product   
8          Soda        77              7     9.0   both  Mid-range product   
9          Soda        77              7     9.0   both  Mid-range product   
10         Soda        77              7     9.0   both  Mid-range product   
11         Soda        77              7     9.0   both  Mid-range product   
12         Soda        77              7     9.0   both  Mid-range product   
13         Soda        77              7     9.0   both  Mid-range product   
14         Soda        77              7     9.0   both  Mid-range product   

       busiest day        busiest_days busiest_period_of_day  max_order  
0   Regularly busy  Regulary busy days           Most orders         10  
1   Regularly busy        Slowest days        Average orders         10  
2   Regularly busy        Slowest days           Most orders         10  
3       Least busy        Slowest days        Average orders         10  
4       Least busy        Slowest days           Most orders         10  
5   Regularly busy  Regulary busy days        Average

In [13]:
# deriving columns with loc()
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [14]:
# printing the freq of column loyality_flag
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

# 6.Task 4.8

In [15]:
# Q2) Find aggregated mean for the entire dataset
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

Q3) Analyse the result above
The mean value differs for the subset and entire dataframe. When checking for a small sample the values are skewed.

In [16]:
# Q4) Creating loyalty flag 
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(20)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10

In [17]:
# Q5) 
ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean', 'min', 'max', 'sum']})

prices                           
                       mean  min      max          sum
loyalty_flag                                          
Loyal customer    10.386336  1.0  99999.0  106814042.2
New customer      13.294670  1.0  99999.0   83011787.2
Regular customer  12.495717  1.0  99999.0  198391693.2

Loyal customers on average spend the least amount of money on their products, whereas new customers spend the most amount of money on their purchases.

In [18]:
# Q6) Different type of spenders
ords_prods_merge['average_price']= ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [19]:
#check accuracy
ords_prods_merge.head(60)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    2539329        1             1          2                  8   
1    2398795        1             2          3                  7   
2     473747        1             3          3                 12   
3    2254736        1             4          4                  7   
4     431534        1             5          4                 15   
5    3367565        1             6          2                  7   
6     550135        1             7          1                  9   
7    3108588        1             8          1                 14   
8    2295261        1             9          1                 16   
9    2550362        1            10          4                  8   
10   2968173       15            15          1                  9   
11   1870022       15            17          2                 16   
12   1911383       15            18          2                 11   
13   2715276       15            21          1                  9   
14    487368       15            22          1                 10   
15   2293453       19             2          5                 14   
16   1973799       19             5          6                 12   
17    532817       19             7          4                 17   
18   1573906       21            10          3                 10   
19   1593000       31            10          3                  8   
20   2231262       31            17          3                 11   
21   2580647       43             6          4                 16   
22   2187180       43             9          4                 12   
23   2497897       52             1          1                  9   
24   1318871       52             2          1                 10   
25   1261384       52             3          1                 10   
26    580568       52             5          2                 10   
27   2428073       52             7          2                  7   
28    342306       52             9          1                 11   
29   1520353       52            10          2                 10   
30   2542086       52            11          2                  9   
31    944694       52            12          1                 16   
32   1268191       52            15          2                 11   
33   1122089       52            17          2                 10   
34   1498922       52            18          2                 10   
35    180919       52            19          3                  8   
36   2695875       52            21          5                 10   
37   3205960       67             1          3                 11   
38    970922       67             2          1                 15   
39    596585       67             3          1                 13   
40   2037087       67             4          2                  9   
41   2474111       67             5          3                 14   
42   1568481       67             6          3                 12   
43   2048527       67             8          2                 11   
44   2389031       67             9          2                 14   
45   2378562       67            10          1                 10   
46   1515698       67            11          1                 11   
47   1928050       67            12          2                 12   
48   1949274       67            13          1                 10   
49    551869       67            14          1                  9   
50    768572       67            16          4                 10   
51    150528       67            18          5                 16   
52     29758       67            19          0                 14   
53   1558976       67            21          1                  9   
54    863974       67            22          1                 11   
55   2308399       67            23          1                 14   
56   2759128       81             3          1                  9   
57

In [20]:
#derive column conditions
ords_prods_merge.loc[ords_prods_merge['average_price']<10, 'spending_flag']= 'Low spender'

In [21]:
ords_prods_merge.loc[ords_prods_merge['average_price']>=10, 'spending_flag']= 'High spender'

In [22]:
#frequency of spending_flag
ords_prods_merge['spending_flag'].value_counts(dropna= False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

In [23]:
ords_prods_merge[['user_id', 'spending_flag', 'loyalty_flag', 'order_number']].head(15)

user_id spending_flag      loyalty_flag  order_number
0         1   Low spender      New customer             1
1         1   Low spender      New customer             2
2         1   Low spender      New customer             3
3         1   Low spender      New customer             4
4         1   Low spender      New customer             5
5         1   Low spender      New customer             6
6         1   Low spender      New customer             7
7         1   Low spender      New customer             8
8         1   Low spender      New customer             9
9         1   Low spender      New customer            10
10       15   Low spender  Regular customer            15
11       15   Low spender  Regular customer            17
12       15   Low spender  Regular customer            18
13       15   Low spender  Regular customer            21
14       15   Low spender  Regular customer            22

In [24]:
# Q7 create order frequency flag
ords_prods_merge['order_frequency_flag']= ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [26]:
ords_prods_merge.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                   999.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  _merge    price_range_loc     busiest day  \
0         Soda  ...    both  Mid-range product  Regularly busy   
1         Soda  ...    both  Mid-range product  Regularly busy   
2         Soda  ...    both  Mid-range product  Regularly busy   
3         Soda  ...    both  Mid-range product      Least busy   
4         Soda  ...    both  Mid-range product      Least busy   

         busiest_days busiest_period_of_day max_order  loyalty_flag  \
0  Regulary busy days           Most orders        10  New customer   
1        Slowest days        Average orders        10  New customer   
2        Slowest days           Most orders        10  New customer   
3        Slowest days        Average orders        10  New customer   
4        Slowest days           Most orders        10  New customer   

  average_price  spending_flag order_frequency_flag  
0      6.367797    Low spender                 21.0  
1      6.367797    Low spender                 21.0  
2      6.367797    Low spender                 21.0  
3      6.367797    Low spender                 21.0  
4      6.367797    Low spender                 21.0  

[5 rows x 23 columns]

In [27]:
#derive column conditions
ords_prods_merge.loc[ords_prods_merge['order_frequency_flag']>20, 'customer_frequency_flag']= 'Non frequent customer'

In [28]:
ords_prods_merge.loc[(ords_prods_merge['order_frequency_flag']>10) & (ords_prods_merge['order_frequency_flag']<=20)  , 'customer_frequency_flag']= 'Regular customer'

In [29]:
ords_prods_merge.loc[ords_prods_merge['order_frequency_flag']<=10, 'customer_frequency_flag']= 'frequent customer'

In [32]:
ords_prods_merge[['user_id', 'loyalty_flag', 'spending_flag', 'customer_frequency_flag']].head(60)

user_id      loyalty_flag spending_flag customer_frequency_flag
0         1      New customer   Low spender   Non frequent customer
1         1      New customer   Low spender   Non frequent customer
2         1      New customer   Low spender   Non frequent customer
3         1      New customer   Low spender   Non frequent customer
4         1      New customer   Low spender   Non frequent customer
5         1      New customer   Low spender   Non frequent customer
6         1      New customer   Low spender   Non frequent customer
7         1      New customer   Low spender   Non frequent customer
8         1      New customer   Low spender   Non frequent customer
9         1      New customer   Low spender   Non frequent customer
10       15  Regular customer   Low spender        Regular customer
11       15  Regular customer   Low spender        Regular customer
12       15  Regular customer   Low spender        Regular customer
13       15  Regular customer   Low spender        Regular customer
14       15  Regular customer   Low spender        Regular customer
15       19      New customer   Low spender       frequent customer
16       19      New customer   Low spender       frequent customer
17       19      New customer   Low spender       frequent customer
18       21  Regular customer   Low spender       frequent customer
19       31  Regular customer   Low spender       frequent customer
20       31  Regular customer   Low spender       frequent customer
21       43  Regular customer   Low spender       frequent customer
22       43  Regular customer   Low spender       frequent customer
23       52  Regular customer   Low spender       frequent customer
24       52  Regular customer   Low spender       frequent customer
25       52  Regular customer   Low spender       frequent customer
26       52  Regular customer   Low spender       frequent customer
27       52  Regular customer   Low spender       frequent customer
28       52  Regular customer   Low spender       frequent customer
29       52  Regular customer   Low spender       frequent customer
30       52  Regular customer   Low spender       frequent customer
31       52  Regular customer   Low spender       frequent customer
32       52  Regular customer   Low spender       frequent customer
33       52  Regular customer   Low spender       frequent customer
34       52  Regular customer   Low spender       frequent customer
35       52  Regular customer   Low spender       frequent customer
36       52  Regular customer   Low spender       frequent customer
37       67  Regular customer   Low spender       frequent customer
38       67  Regular customer   Low spender       frequent customer
39       67  Regular customer   Low spender       frequent customer
40       67  Regular customer   Low spender       frequent customer
41       67  Regular customer   Low spender       frequent customer
42       67  Regular customer   Low spender       frequent customer
43       67  Regular customer   Low spender       frequent customer
44       67  Regular customer   Low spender       frequent customer
45       67  Regular customer   Low spender       frequent customer
46       67  Regular customer   Low spender       frequent customer
47       67  Regular customer   Low spender       frequent customer
48       67  Regular customer   Low spender       frequent customer
49       67  Regular customer   Low spender       frequent customer
50       67  Regular customer   Low spender       frequent customer
51       67  Regular customer   Low spender       frequent customer
52       67  Regular customer   Low spender       frequent customer
53       67  Regular customer   Low spender       frequent customer
54       67  Regular customer   Low spender       frequent customer
55       67  Regular customer   Low spender       frequent customer
56       81      New customer   Low spender   Non frequent customer
57       81      New customer   Low spender   Non frequent c

I can see that user_id 19 is also considered frequent customer

In [37]:
# Investigating why the user_id 19 has different results for the loyalty flag & customer frequency flag
investigate=ords_prods_merge[ords_prods_merge.user_id == 19]

In [51]:
investigate.head(60)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
15         2293453       19             2          5                 14   
16         1973799       19             5          6                 12   
17          532817       19             7          4                 17   
2269387    2474845       19             6          5                 13   
3028920      86918       19             9          5                 16   
5101635    2293453       19             2          5                 14   
5101636    1973799       19             5          6                 12   
8116470     532817       19             7          4                 17   
8479132    2293453       19             2          5                 14   
9210215    2483716       19             8          4                  4   
9314597    2293453       19             2          5                 14   
9315663     532817       19             7          4                 17   
9430793      86918       19             9          5                 16   
10058025   2808909       19             4          5                 12   
10078217    532817       19             7          4                 17   
10199285   1321974       19             1          6                 13   
10199286   2293453       19             2          5                 14   
10200896   1321974       19             1          6                 13   
10200897   2293453       19             2          5                 14   
10200898   2808909       19             4          5                 12   
10200899   1973799       19             5          6                 12   
10201096   1321974       19             1          6                 13   
10201097   2808909       19             4          5                 12   
10201892   1321974       19             1          6                 13   
10201980   1321974       19             1          6                 13   
10202033   1321974       19             1          6                 13   
10202151   1321974       19             1          6                 13   
10202152   2474845       19             6          5                 13   
10202168   1321974       19             1          6                 13   
10202169   2474845       19             6          5                 13   
10202173   1321974       19             1          6                 13   
10202246   1321974       19             1          6                 13   
10202270   1321974       19             1          6                 13   
10202341   1321974       19             1          6                 13   
10202374   1321974       19             1          6                 13   
10202416   1321974       19             1          6                 13   
10202452   1321974       19             1          6                 13   
10202453   2293453       19             2          5                 14   
10202454   2208892       19             3          0                 14   
10202455   1973799       19             5          6                 12   
10202456   2474845       19             6          5                 13   
10202457    532817       19             7          4                 17   
10202458   2483716       19             8          4                  4   
10202459     86918       19             9          5                 16   
10209569   1321974       19             1          6                 13   
10209570   2293453       19             2          5                 14   
10209571   2808909       19             4          5                 12   
10211665   1321974       19             1          6                 13   
10211666   2293453       19             2          5                 14   
10211667   2808909       19             4          5                 12   
10212714   1321974       19             1          6                 13   
10212715   2293453       19             2          5                 14   
10212716   2208892       19        

In [39]:
investigate['days_since_prior_order'].value_counts(dropna= False)

6.0      69
8.0      46
999.0    32
5.0      30
20.0     16
16.0      6
7.0       5
Name: days_since_prior_order, dtype: int64

We can see from the above result that 32 number of 999 values are there(missing values represented as 999). The order_frequency column(median value) for the 999 values have been assigned 7. This maybe the cause of different flags. Hence, the customer_frequency_flag need to be cross checked before using it for further analysis as the missing values might affect the result. 

# 7. Export data

In [52]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_3.pkl'))